### PR TITLE
8316141: Improve CEN header validation checking

### PR DESCRIPTION
**Pull request by LanceAndersen**
Please review this  PR which enhances the existing CEN header validation checking to ensure that the
 size of the CEN Header + name length + comment length + extra length do not exceed 65,535 bytes per the PKWare APP.NOTE 4.4.10, 4.4.11, & 4.4.12. Also check that current CEN header will not exceed the length of the CEN array.

Mach 5 tiers 1-3 are clean with this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316141](https://bugs.openjdk.org/browse/JDK-8316141): Improve CEN header validation checking (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16570/head:pull/16570` \
`$ git checkout pull/16570`

Update a local copy of the PR: \
`$ git checkout pull/16570` \
`$ git pull https://git.openjdk.org/jdk.git pull/16570/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16570`

View PR using the GUI difftool: \
`$ git pr show -t 16570`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16570.diff">https://git.openjdk.org/jdk/pull/16570.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16570#issuecomment-1802572011)

### CODE REVIEW

**Comment by eirbjo** [comment]
Perhaps the PR/issue title could be more specific in describing what is being validated? Something like "Validate the combined length of CEN header fields"?

File: src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java
Comment (eirbjo) [inline src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java:1596]: If the corresponding ZipFile local variable is renamed, this should also be updated.

File: src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java
Comment (eirbjo) [inline src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java:1668]: I think my ZipFile comment also applies here.

**Review by eirbjo — COMMENTED** [review]
LGTM, although I'm not sure I follow the underlying motivation of this stricter validation. 

What problem is being solved here? APPNOTE.TXT uses the phrase `SHOULD NOT`. Even if the spec is not an RFC, RFC2119 defines `SHOULD NOT` as:

```This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.
```

I would expect our producer `ZipOutputStream` to be stricter than our consumers in this case, honoring Postel's law. From a implementation robustness perspective, the individual lengths are already validated, it's just the combined clause that is now enforced in this PR.

That said, here are some comments inline:

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:1235]: `entryPos` was the name of the field from a previous PR so I did not see a need to change it and decided there was no need to keep extraStartingOffset given the change in validation above.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:1225]: I can remove the cast as that was a holdover.  I chose to  make this a long knowing that it would not overflow but  an overflow while unlikely could occur depending on the value of `pos`   in the statement below

```
if (headerSize > 0xFFFF || pos + headerSize > cen.length - ENDHDR) {
                zerror("invalid CEN header (bad header size)");
 }
```

I could keep headerSize an int and then move the cast to the if statement:

```
if (headerSize > 0xFFFF || (long)pos + headerSize > cen.length - ENDHDR) {
                zerror("invalid CEN header (bad header size)");
 }
```
I  decided making headerSize a long  might be clearer but do not have a strong preference and will go with the consensus 

As far as the name,  I don't have a strong preference, but not sure combinedLength is any better

**Review by LanceAndersen — COMMENTED** [review]
Thank you for the comments.  See my replies below.

Regarding you comment about checking  whether or not to check if the combined length of the CEN header + name length + comment length + extra length > 65K bytes,  I chose to add this given the strong wording given this a really old spec.  That being said, I do not object to removing the validation if that is the overall preference.
                zerror("invalid CEN header (bad header size)");
            }

**Comment by eirbjo** [comment]
> Regarding you comment about checking whether or not to check if the combined length of the CEN header + name length + comment length + extra length > 65K bytes, I chose to add this given the strong wording given this is a really old spec. That being said, I do not object to removing the validation if that is the overall preference.

I can't claim to have a particularly strong opinion on this, the following is mostly me thinking aloud:

- Given Hyrum's Law, it is conceivable that someone is currently using the extra or comment fields to attach up to 65535+65535 bytes of metadata for entires. The proposed validation will break such schemes. Does Oracle have a ZIP file corpus which could be used to identify files currently exceeding the combined length clause, just to get a sense of how rare or common this is?
- The actual benefits of adding this validation after all these years is not quite clear to me. I don't see how this improves security, robustness, compatibility, maintainability or other ilities (apart from strictly-following-the-spec-ility :-)
- I created a ZIP file with an entry with an extra field of the maximal expressable length of 0xFFFF. Info-ZIP's `unzip` command on MacOS did not output any warning or error when processing this file.

**Comment by LanceAndersen** [comment]
> > Regarding you comment about checking whether or not to check if the combined length of the CEN header + name length + comment length + extra length > 65K bytes, I chose to add this given the strong wording given this is a really old spec. That being said, I do not object to removing the validation if that is the overall preference.
> 
> I can't claim to have a particularly strong opinion on this, the following is mostly me thinking aloud:
> 
> * Given Hyrum's Law, it is conceivable that someone is currently using the extra or comment fields to attach up to 65535+65535 bytes of metadata for entires. The proposed validation will break such schemes. Does Oracle have a ZIP file corpus which could be used to identify files currently exceeding the combined length clause, just to get a sense of how rare or common this is?
> * The actual benefits of adding this validation after all these years is not quite clear to me. I don't see how this improves security, robustness, compatibility, maintainability or other ilities (apart from strictly-following-the-spec-ility :-)
> * I created a ZIP file with an entry with an extra field of the maximal expressable length of 0xFFFF. Info-ZIP's `unzip` command on MacOS did not output any warning or error when processing this file.

Yes we have a corpus search available and have exercised this patch (along with your ZipInputStream patch) without any regressions.

Given where we are in the JDK 22 cycle, going to hold off on finalizing the PR until we fork for JDK 23 and look to move this forward early on allowing for additional time to bake

**Comment by AlanBateman** [comment]
> Given where we are in the JDK 22 cycle, going to hold off on finalizing the PR until we fork for JDK 23 and look to move this forward early on allowing for additional time to bake

Tightening validation always comes with risk. Doing it early in JDK 23 to allow time for course correction if needed seems a good plan.

**Comment by eirbjo** [comment]
> Doing it early in JDK 23 to allow time for course correction if needed seems a good plan.

Another benefit is that if we should decide to validate LOC headers similarly in `ZipInputStream`, delaying until 23 will allow us to introduce these very similar changes in the same release.

**Comment by eirbjo** [comment]
While investigating an unrelated issue, I noticed that Android's `zipalign` tool processes zip files and injects data into the extra field  to make the beginning of the file data be word-aligned or page-aligned:

**Review by AlanBateman — APPROVED** [review]
I think the zip changes are okay. As per our discussion here, the compatibility impact can be evaluated later in JDK 23 to gauge whether there it is too strict.

**Pull request closed without merging**
